File: src/py/agent_studio/memory/export.py
"""Render a conversation to Markdown (parity with the TS exporter)."""

from __future__ import annotations

import re

from .store import Conversation

_ROLE_HEADING = {"user": "### You", "assistant": "### Assistant", "system": "### System"}


def conversation_to_markdown(conversation: Conversation) -> str:
    lines: list[str] = [f"# {conversation.title}", ""]
    if conversation.model:
        lines.append(f"- Model: {conversation.model}")
    if conversation.provider_id:
        lines.append(f"- Provider: {conversation.provider_id}")
    lines.append(f"- Created: {conversation.created_at}")
    lines.append(f"- Updated: {conversation.updated_at}")
    lines.append("")
    for m in conversation.messages:
        lines.append(_ROLE_HEADING.get(m.role, f"### {m.role}"))
        lines.append("")
        lines.append(m.content)
        lines.append("")
    return re.sub(r"\n+$", "\n", "\n".join(lines))


def default_export_filename(conversation: Conversation) -> str:
    source = conversation.title if conversation.title and conversation.title != "New conversation" else "conversation"
    slug = re.sub(r"[^a-z0-9]+", "-", source.lower()).strip("-")[:40] or "conversation"
    return f"{slug}-{conversation.id[:8]}.md"

File: src/engine/memory/export.ts
// Render a conversation to portable Markdown so users can drop results into
// docs, PRs, or notes. Pure and dependency-free.

import type { Conversation } from './types.ts';

const ROLE_HEADING: Record<string, string> = {
  user: '### You',
  assistant: '### Assistant',
  system: '### System',
};

/** Render a conversation as a Markdown document. */
export function conversationToMarkdown(conversation: Conversation): string {
  const lines: string[] = [];
  lines.push(`# ${conversation.title}`);
  lines.push('');
  const meta: string[] = [];
  if (conversation.model) meta.push(`- Model: ${conversation.model}`);
  if (conversation.providerId) meta.push(`- Provider: ${conversation.providerId}`);
  meta.push(`- Created: ${conversation.createdAt}`);
  meta.push(`- Updated: ${conversation.updatedAt}`);
  lines.push(...meta);
  lines.push('');

  for (const m of conversation.messages) {
    lines.push(ROLE_HEADING[m.role] ?? `### ${m.role}`);
    lines.push('');
    lines.push(m.content);
    lines.push('');
  }
  return lines.join('\n').replace(/\n+$/, '\n');
}

/** A filesystem-safe default export filename for a conversation. */
export function defaultExportFilename(conversation: Conversation): string {
  const slugSource = conversation.title && conversation.title !== 'New conversation' ? conversation.title : 'conversation';
  const slug = slugSource
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
    .slice(0, 40) || 'conversation';
  return `${slug}-${conversation.id.slice(0, 8)}.md`;
}
